src/libstd/sys/unix/process.rs

```diff
@@ -13,37 +13,157 @@
 use prelude::v1::*;
 use os::unix::prelude::*;
 
+use borrow::Borrow;
 use collections::HashMap;
 use env;
 use ffi::{OsString, OsStr, CString, CStr};
 use fmt;
+use hash::{Hash, Hasher};
 use io::{self, Error, ErrorKind};
+use iter;
 use libc::{self, pid_t, c_void, c_int, gid_t, uid_t};
 use ptr;
+use slice;
 use sys::fd::FileDesc;
 use sys::fs::{File, OpenOptions};
 use sys::pipe::AnonPipe;
 use sys::{self, cvt, cvt_r};
 
 ////////////////////////////////////////////////////////////////////////////////
+// ValidatedCString
+////////////////////////////////////////////////////////////////////////////////
+
+// A `ValidatedCString` lets us accept an `OsString` that is intended to
+// be passed as a C string, and report an error at the time it is used
+// as one.
+//
+// For `std::process;:Command`, this would look like
+//
+//     let child = Command::new("has\0nuls\0inside")
+//                         .arg("nuls\0here\0too")
+//                         .cwd("and\0here")
+//                         .env("and\0even", "in\0here")
+//                         .spawn();
+//
+// The `std::process` API only allows reporting failure at `spawn()` and
+// its equivalents; all the builder style methods others return `'&mut
+// self` for chaining. The `ValidatedCString` lets us delay the error reporting
+// to that point and report an error via a `Result` rather than by
+// panicking.
+#[derive(Debug, Clone)]
+pub enum ValidatedCString {
```

Comment:
Maybe (ab)use `Option` for this?

```
pub struct Command {
    program: Option<CString>,
    args: Vec<Optrion<ValidatedCString>>,
    env: Option<HashMap<Option<CString>, Option<ValidatedCString>>>,
    cwd: Option<Option<CString>>,
    uid: Option<uid_t>,
    gid: Option<gid_t>,
    session_leader: bool,
    saw_nul: bool,
}
```

The `Debug` impl could then use `unwrap_or("<string-with-nul>")`.

This is a bit weird for env, as the hash map would lose context of how many vars were nulful. The cwd would be a bit unwieldy with the double `Option`...